### ISSUE TITLE
Having some trouble with Typescript and neomake

### ISSUE THREAD

**Issue by JoosepAlviste** [janko/vim-test#382]
I can't get `neomake` to show errors in Typescript files. The test results work fine but the virtual text and gutter signs do not show up. Also, it works in `javascript` files. I printed out the `g:neomake_hook_context.jobinfo` but it looks fine to me. Looking at `neomake`s logs, it seems like `neomake` doesn't detect the buffer number correctly.

I used this starter template for my typescript project: https://github.com/alexjoverm/typescript-library-starter.

Any ideas what else I could do to debug this?

Here's the relevant parts in my `vimrc` (shouldn't be anything special, right?):

```vim
let g:test#strategy = 'neomake'
let g:test#javascript#jest#options = '--reporters ' . $HOME . '/.config/yarn/global/node_modules/jest-vim-reporter'
let g:neomake_warning_sign = {
  \   'text': '◉'
  \ }
let g:neomake_error_sign = {
  \   'text': '◉'
  \ }

let g:neomake_open_list = 1
```

Here is the `g:neomake_hook_context.jobinfo`:

```
{'nvim_job': 16, 'cwd': '<myProjectRoot>', 'file_mode': 0, 'as_string': function('317'), 'cd_back_cmd': '', 'bufnr': 2, 'name'
: 'makeprg', 'maker': {'_get_fname_for_args': function('290'), '__command_is_string': 1, '_bind_args': function('295'), '_get_tempfilename': function('
292'), 'name': 'makeprg', 'auto_enabled': 1, 'exe': '/usr/local/bin/zsh', '_get_fname_for_buffer': function('294'), 'remove_invalid_entries': 0, '_get_
default_tempfilename': function('293'), 'append_file': 0, 'buffer_output': 1, 'output_stream': 'both', '_get_argv': function('297'), 'errorformat': '%+
A %#%f %#(%l\,%c): %m,%C%m', 'args': ['-c', 'node_modules/.bin/jest --reporters node_modules/jest-vim-reporter --no-c
overage -- src/file.test.ts']}, 'pending_output': [], 'jobstart_opts': {'stderr': [''], 'stdout_buffered': 1, 'on_exit': function('<SNR>133
_nvim_exit_handler_buffered'), 'stderr_buffered': 1, 'stdout': ['Dummy test works if true is truthy', '<myProjectRoot>/src/file.test.ts:8:18: Error: expect(received).toBeFalsy() Received: true', '']}, 'get_pid': function('316'), 'make_id': 4, 'id': 4, 'serialize':
0, 'finished': 1, 'cd_back': function('318'), 'filter_output': '', 'buffer_output': 1, 'output_stream': 'both', 'exit_code': 1, 'ft': 'typescript', 'cd
': function('319'), 'argv': ['/usr/local/bin/zsh', '-c', 'node_modules/.bin/jest --reporters /node_modules/jest-vim-re
porter --no-coverage -- src/file.test.ts']}
```

And here's the log for `neomake` (note `Skipped 2 entries without bufnr` and `bufnr` is `0` for the error):

```
13:04:54 10408 [D +59.7] [7.-:2:1] Running makers: makeprg (auto).
13:04:54 10408 [D      ] [-.-:2:1] Using setting append_file=0 from 'maker'.
13:04:54 10408 [V      ] [7.7:2:1] Starting async job: /usr/local/bin/zsh -c 'node_modules/.bin/jest --reporters node_modules/jest-vim-reporter --no-coverage -- src/vim-ts-test-test.test.ts'.
13:04:54 10408 [D      ] [7.7:2:1] cwd: <myProjectRoot>.
13:04:54 10408 [V      ] [7.7:2:1] Calling User autocmd NeomakeJobStarted with context: {'jobinfo': 'Job 7: makeprg'}.
13:04:56 10408 [D +1.34] [7.7:2:1] output on stdout: ['Dummy test works if true is truthy', '<myProjectRoot>/src/file.test.ts:8:18: Error: expect(received).toBeFalsy() Received: true', ''].
13:04:56 10408 [D      ] [7.7:2:1] exit: makeprg: 1.
13:04:56 10408 [D      ] [7.7:2:1] Processing 2 lines of output.
13:04:56 10408 [D      ] [7.-:2:1] Adding 2 list entries.
13:04:56 10408 [D      ] [7.-:2:1] Creating quickfix list for entries.
13:04:56 10408 [D      ] list: call: set: [[], ' ', {'context': {'neomake': {'make_info': 'make_id=7'}}, 'items': [], 'title': 'Neomake[project]: makeprg...'}].
13:04:56 10408 [D      ] list: got qfid (action=set): 26.
13:04:56 10408 [D      ] list: call: set: [[], 'a', {'id': 26, 'items': [{'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol': 0, 'nr': -1, 'module': '', 'type': '', 'pattern': '', 'text': 'Dummy test works if true is truthy'}, {'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol': 0, 'nr': -1, 'module': '', 'type': '', 'pattern': '', 'text': '<myProjectRoot>/src/file.test.ts:8:18: Error: expect(received).toBeFalsy() Received: true'}], 'title': 'Neomake[project]: makeprg...'}].
13:04:56 10408 [D      ] list: call: "get", returning items: [['getqflist', [{'id': 26, 'items': 1}]]].
13:04:56 10408 [D      ] [7.7:2:1] Processing 2 entries.
13:04:56 10408 [D      ] [7.-:2:1] Project-level errors cleaned.
13:04:56 10408 [D      ] [7.7:2:1] Skipped 2 entries without bufnr: [{'lnum': 0, 'bufnr': 0, 'col': 0, 'pattern': '', 'valid': 0, 'vcol': 0, 'nr': -1, 'type': '', 'module': '', 'text': 'Dummy test works if true is truthy'}, {'lnum': 0, 'bufnr': 0, 'col': 0, 'pattern': '', 'valid': 0, 'vcol': 0, 'nr': -1, 'type': '', 'module': '', 'text': '<myProjectRoot>/src/file.test.ts:8:18: Error: expect(received).toBeFalsy() Received: true'}].
13:04:56 10408 [D      ] [7.7:2:1] Handling quickfix list: executing cwindow.
13:04:56 10408 [D      ] [7.7:2:1] makeprg: completed with exit code 1.
13:04:56 10408 [D      ] [7.7:2:1] Cleaning jobinfo.
13:04:56 10408 [V      ] [7.7:2:1] Calling User autocmd NeomakeJobFinished with context: {'jobinfo': 'Job 7: makeprg [finished]'}.
13:04:57 10408 [D +1.63] Removing signs: {file: {}, project: {}}.
13:04:57 10408 [D      ] list: call: "title": [[], 'a', {'id': 26, 'title': 'Neomake[project]: makeprg(2)'}].
13:04:57 10408 [D      ] [7.-:2:1] Cleaning make info.
```

**Comment by rodrigore** [comment]
Same situation here. When I running PHPunit with Neomake strategy I don't see any symbols or signr.

**Comment by janko** [comment]
Does it work when you run Neomake directly? All vim-test does is run the `:NeomakeProject` command (with [automatic compiler setup](https://github.com/janko/vim-test/blob/5aaa872eb7519909b87abb2b0f13600aa2210771/autoload/test/strategy.vim#L112-L137)).  What would be the correct Neomake command to run in order to get the desired results?

**Comment by rodrigore** [comment]
I'm using neovim with neomake hooks, and putting on my status bar the result. 
```
function!` MyOnNeomakeJobFinished() abort
  let context = g:neomake_hook_context
  if context.jobinfo.exit_code == 0
    let g:TESTING_STATUS = 'passing'
  endif
  if context.jobinfo.exit_code == 1
    let g:TESTING_STATUS = 'failing'
  endif
endfunction

augroup my_neomake_hooks
  au!
  autocmd User NeomakeJobFinished call MyOnNeomakeJobFinished()
  autocmd User NeomakeJobStarted call MyOnNeomakeJobStarted()
augroup END
```
I wish to get some feedback other than using the quickfix but I'm not sure if the error and warning sign should be appear.

**Comment by janko** [comment]
When you use Neomake directly, to these hooks work as expected? I'm trying to determine whether the problem is in Neomake itself, or in how vim-test calls Neomake.

**Comment by rodrigore** [comment]
Im not sure if I undertands what do you mean as expected, but searching on neomake repository I noticed that inside php configuration it doesn't exist *phpunit* so I assume that the answer of your answer is that Neomake doesn't format the result of phpunit but maybe I'm wrong.
https://github.com/neomake/neomake/blob/master/autoload/neomake/makers/ft/php.vim

If this doesn't respond your answer can you give me the example of using Neomake directly?

The hooks that I use only serve for show  in the status bar the result of the test. For example I have this test that failed on line 15 :
![image](https://user-images.githubusercontent.com/1557305/64586076-805f3880-d371-11e9-8f27-09c00b0fe558.png)

I wish that some way get the output of neomake to know that in that line was the error, but maybe that is not possible.

For now my only option is to use the  Quickfix.
![image](https://user-images.githubusercontent.com/1557305/64586173-d92ed100-d371-11e9-93ed-cbf5514bfc06.png)

**Comment by janko** [comment]
> If this doesn't respond your answer can you give me the example of using Neomake directly?

Neomake's docs have the usage. The main commands seem to be `:Neomake` and `:Neomake!`.

The reason I'm asking is because I'm not a Neomake expert, and don't use it myself. So I want someone else to figure out what's the correct Neomake command to run, and then I can change `vim-test` to use it.

I don't want that the integration between vim-test and Neomake is treated as some sort of a black box. vim-test generates a shell command to run, and passes it to Neomake as can be seen in `autoload/test/strategy.vim`. If this does the wrong thing, we need to figure out how to call Neomake so that it does the right thing.

However, I don't have enough free time anymore to figure this out by myself, that's why I'm asking for help.

> searching on neomake repository I noticed that inside php configuration it doesn't exist phpunit so I assume that the answer of your answer is that Neomake doesn't format the result of phpunit but maybe I'm wrong.

vim-test doesn't use Neomake's "makers", so that might be the issue. It uses Vim's compilers, finding them with dispatch.vim. Did Neomake start to reimplement Vim's compilers? Seems like it would lead to a lot of duplicated effort.

**Comment by rodrigore** [comment]
Im also not Neomake expert, in fact only yesterday I installed only to make pretty my status bar.

Anyway, searching another Dotfiles on Github I found some neomake 'maker' which format the output of phpunit and it work for display the symbol errors.

![image](https://user-images.githubusercontent.com/1557305/64624408-ffc82880-d3c0-11e9-87c7-d1b64077e635.png)

The source of the phpunit maker is this: https://github.com/laserswald/dotfiles/blob/2360447f3204aaf87e409eb1a0836aebe4ce9ec8/vim/.vim/plugin/pkg/neomake.vim

The way to executed is: 
`:Neomake! phpunit`

Now I'm not sure how can I use that with vim-test, but I will keep trying.

**Comment by janko** [comment]
Ok, this really helps. Now, it would be ideal if Neomake had some kind of automatic maker selection based on the shell command prefix, something like [`g:dispatch_compilers`](https://github.com/tpope/vim-dispatch/blob/a76bec9196fe27e195d167a5c2ee1da763d31b96/doc/dispatch.txt#L53-L61) from dispatch.vim.

That way vim-test wouldn't need to know the appropriate maker for a shell command, it would be automatically selected. Otherwise I'm not sure how vim-test could determine the maker itself, perhaps with a similar config like `g:dispatch_compilers`. However, I would prefer if this variable lived inside Neomake, so that vim-test has less to do.

**Comment by janko** [comment]
Hmm, wait, maybe `:NeomakeSh` would work, it seems to have some automatic detection (it's not documented, but I see something in the source code). Could you try whether running

```
:NeomakeSh phpunit <some more args>
```

will select the `phpunit` maker and correctly parse results?

If that works, I guess it would make sense for vim-test to switch to that. However, that means if anyone is relying on Vim compilers for which Neomake doesn't yet have makers, result parsing will not work for them. However, considering that Neomake seems to have drifted away from Vim compilers, maybe that's ok.

**Comment by rodrigore** [comment]
I try with `:NeomakeSh ./vendor/bin/phpunit` and didn't show the symbols, it works just like the actual situation.

I noticed that if I run: `:NeomakeProject phpunit` It work fine just like: `:Neomake! phpunit`

In this case phpunit is the name of my custom maker:
```
let g:neomake_php_phpunit_maker = {
            \ 'exe': './vendor/bin/phpunit',
            \ 'process_output': function("PhpUnitProcessOutput"),
            \ }
```
Reading the docs of Neomake I saw that is possible override the makers of a custom language, something like this: `let g:neomake_php_maker = ['phpunit']` and suppose when I run `:Neomake!` should be use the makers inside of that array, but didn't work for me.

**Comment by rodrigore** [comment]
Hi Again, I was trying to use the neomake maker with vim-test and I dont find the solution, so I was reading the source code of vim-test I noticed that I can define some errorformat (efm) and the result with that is that now I got the warning message and warning sign 😄 

I write a simple errorformat for php, which is not perfect but only for show the warning signs and messages when using phpunit.

`errorformat=%f:%l`

**Comment by JoosepAlviste** [comment]
I think I also got my problem solved -- I think the problem ended up being the `errorformat` just as @rodrigore mentioned. So, I just set the following `errorformat` in `ftplugin/typescript.vim`:

```vim
setlocal errorformat=%f:%l:%c:\ %m
```

Now Neomake can parse the error correctly from the `jest-vim-reporter` output.

I think that this issue can probably be closed now since I figured it out. Also, this might have been more related to Neomake than vim-test.
